features/storage/kvstore/kv_map/KVMap.h

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2018 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef _KV_MAP
+#define _KV_MAP
+
+#include "KVStore.h"
+#include "platform/PlatformMutex.h"
+#include "platform/SingletonPtr.h"
+#include "BlockDevice.h"
+#include "FileSystem.h"
+
+namespace mbed {
+
+#define  MAX_ATTACHED_KVS 3
+
+typedef struct {
+    KVStore *kvstore_main_instance;
+    KVStore *internal_store;
+    KVStore *external_store;
+    BlockDevice *internal_bd;
+    BlockDevice *external_bd;
+    FileSystem *external_fs;
+} kvstore_config_t;
+
+typedef struct {
+    char *partition_name;
+    kvstore_config_t *kv_config;
+} kv_map_entry_t;
+
+class KVMap : private mbed::NonCopyable<KVMap> {
+public:
+
+    /**
+     * @brief As a singleton, return the single instance of the class.
+     *        This class is a singleton for the following reasons:
+     *        - Ease of use, so you don't have to coordinate instantiations.
+     *        - Lazy instantiation of internal data, (which we can't achieve with simple static classes).
```

Comment:
Query: Is "lazy instantiation" a common term I haven't heard of, or should we rephrase this?

Reply:
It's a common term.